end_entity: add verify_private_key function.

cpu opened the pull request:
This branch introduces a function on the `EndEntityCert` type that can be used to check if a DER encoded PKCS8 private key matches the EE certificate's subject public key. This resolves the user request made in https://github.com/rustls/rustls/issues/618.

This is done by un-marshalling the encoded subject public key information (SPKI) of the EE certificate and performing a byte-for-byte comparison of the encoded SPKI public key with the encoded public key derived from the private key. If the two match, we consider the private key usable for the certificate.

Future work is available to support ECDSA EC1 encoded private keys, or raw DER encoded RSA private keys, in addition to their PKCS8 encoded counterparts. To keep things simple to start I've only implemented support for PKCS8.

_Note to reviewers: This will eventually have some `generate.py` conflicts with my other open PRs that I will sort out based on which get merged first._

cpu commented on src/signed_data.rs:
```diff
@@ -165,16 +165,16 @@ pub(crate) fn verify_signature(
-    algorithm_id_value: untrusted::Input<'a>,
-    key_value: untrusted::Input<'a>,
+pub(crate) struct SubjectPublicKeyInfo<'a> {
+    pub(crate) algorithm_id_value: untrusted::Input<'a>,
```
I made the `algorithm_id_value` field crate-visible because it felt like it could be useful down the road, but this branch doesn't use this field (_there's no algorithm ID available when we unmarshal a PKCS8 private key_).

If preferred, we could leave this field private and only change the visibility of the top-level struct + `key_value`.

djc replied:
I think we should stick with the minimum visibility now and widen the visibility as we need it.

cpu replied:
SGTM. Reverted that visibility change.

djc commented on src/end_entity.rs:
```diff
@@ -185,4 +193,52 @@ impl<'a> EndEntityCert<'a> {
+    /// If the provided key isn't usable for this EE certificate [`Error::CertPrivateKeyMismatch`]
+    /// will be returned.
+    ///
+    /// Private key encodings other than DER PKCS8 are not supported at this time.
```
Haven't looked at this in much detail yet, but I feel like the docstring here should be very explicit about what key types are supported, and ideally also make it explicit which key types that rustls supports it doesn't support.

(I feel a little hesitant about merging this without support for all the key types that rustls otherwise supports...)

cpu replied:
That's fair, I'll set this as a draft until I have a chance to implement the EC1 and raw DER RSA encoding support. I don't think it will be too much work.

cpu replied:
PKCS1 for RSA was easy. Getting the EC1 support in took some more work. Will call out some points of discussion in some self-review comments.

cpu commented on src/der.rs:
```diff
@@ -222,3 +226,163 @@ macro_rules! oid {
 }
+
+#[cfg(feature = "alloc")]
+pub(crate) fn wrap_in_asn1_len(bytes: &mut Vec<u8>) {
```
To support RFC 5915 Sec1 encoded P-256/P-384 keys this file is now updated with some helpers taken from Rustls, converted to avoid using anything Rustls-specific

* https://github.com/rustls/rustls/blob/main/rustls/src/x509.rs
* https://github.com/rustls/rustls/blob/4d710591c134ee1dbc8ed9715354f3e29709beec/rustls/src/sign.rs#L289-L336

In a perfect world all of this code would go into Ring (see https://github.com/briansmith/ring/pull/1456), but I don't think that will happen anytime soon.

Open questions:
* The more aggressive `clippy` config used by `webpki` required me to allow some `as_conversions`. I've annotated these and indicated why I think they're safe here but would appreciate a double check. Otherwise we should convert these to use `try_Into` and make the functions falliable.

* I haven't exported these helpers for use outside of webpki. I can do that, but so far the webpki `der` module has been crate private and so I wasn't sure if everyone is on-board with this visibility change. It makes sense to me that we'd want to remove the duplicate code from Rustls to use the webpki versions, so perhaps we should make the visbillity change?

djc replied:
This function as specified is a little ugly (I know this the formulation used in rustls today -- tried to optimize it a bit recently and decided to postpone that) because it has to effectively copy the data once for every byte needed to represent the length. It would be nice to reformulate `wrap_in_sequence()`/`wrap_in_asn1_len()` to do something more efficient.

(Ideally if we know the length of the data separately from having access to the buffer the full sequence is about to be encoded into, we would write the sequence tag, the length bytes, and then the data. If the target buffer is the one the data currently lives in, ideally we'd only copy the data once instead of 2 + log n times. For both of these cases, we could make sure that we don't reallocate multiple times while growing the destination buffer.)

(Also as previously discussed, I would usually put these little helper functions below the more substantial, higher-level, more public API in the module.)

cpu replied:
Is https://github.com/rustls/webpki/commit/b8d82e583184d38790364787ae898e87490e67f0 any closer to what you had in mind? If so I'll fix it up and fold back into this branch.

cpu replied:
@djc How about something like https://github.com/rustls/webpki/commit/fb0cc1424c03b49afc913975e0919e4a79e5be84 

Two main differences;
* I'm allocating one Vec up-front that should hold the entire result without needing to expand itself.
* To avoid shifting the data around, I'm building the Vec contents by appending data into the Vec and then flipping it all at the end.

It's a little bit fiddly working with variable length DER encodings and I'm not sure my very specialized "write a DER length but in reverse" helper is up to snuff, but I haven't been able to think of another way to achieve a result that avoids shifting the data around or allocating more than appropriate. 

I might need to pair with you on a different approach if this version isn't workable (or maybe we could land the feature with the Rustls formulation and then iterate on optimization separately?)

LMK what you think :-)

cpu commented on tests/generate.py:
```diff
@@ -810,9 +823,137 @@ def client_auth() -> None:
+            _generate_testcert(name, privkey, serialization_format)
+            _write_test(name)
+
+        # pyca cryptography doesn't have support for serializing SEC1 format EC keys, so we
+        # pre-generate a key/cert with OpenSSL for this case:
```
It might be my own inexperience with pyca cryptography but I wasn't able to find a way to easily marshal an EC private key in SEC 1 format. It looks like there's support for unmarshalling public keys in this format, but not marshalling private keys.

I thought an OK compromise was to generate these ~4 test files by hand with OpenSSL leaving a comment here to reproduce the work if needed.

djc replied:
@alex got any hints on this?

alex replied:
Pretty sure you want https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization/#cryptography.hazmat.primitives.serialization.PrivateFormat.TraditionalOpenSSL as the serialization format to get SEC1. (The name is a bit stupid, I can't remember why)

cpu replied:
Ohhhhh! Yes that did the trick. Thank you :-)

cpu commented:
> cpu force-pushed the cpu-rustls-618-verify-key branch from 9c17609 to 4db6f39 

* Replaced the manually generated SEC1 test cases w/ pyca cryptography generated cases (thanks to Alex for the pointer!)
* Refactored the code soup that I had in `verify_private_key` previously to use smaller helper functions for the key parsing. I think it reads better now :crossed_fingers:

cpu commented:
> cpu force-pushed the cpu-rustls-618-verify-key branch from 4db6f39 to 0507c2f 

Rebased to resolve conflicts.

djc commented on src/end_entity.rs:
```diff
@@ -185,4 +194,86 @@ impl<'a> EndEntityCert<'a> {
+
+    #[cfg(feature = "alloc")]
+    fn ecdsa_from_sec1(sec1_der: &[u8]) -> Option<EcdsaKeyPair> {
+        fn try_sec1_curve(
+            sigalg: &'static EcdsaSigningAlgorithm,
+            sec1_der: &[u8],
+        ) -> Option<EcdsaKeyPair> {
+            der::convert_sec1_to_pkcs8(sigalg, sec1_der)
+                .and_then(|pkcs8_der| {
+                    EcdsaKeyPair::from_pkcs8(sigalg, &pkcs8_der).map_err(|_| Error::BadDer)
+                })
+                .ok()
+        }
```
I think this would be easier to follow if you extracted the nested function.

cpu replied:
Done.

djc commented on src/end_entity.rs:
```diff
@@ -185,4 +194,86 @@ impl<'a> EndEntityCert<'a> {
+        let spki = parse_spki_value(self.inner.spki.value())?;
+        let pubkey = spki.key_value.as_slice_less_safe();
+
+        fn matches_or_error(pubkey: &[u8], pubkey_tbv: &[u8]) -> Result<(), Error> {
+            match pubkey == pubkey_tbv {
+                true => Ok(()),
+                false => Err(Error::CertPrivateKeyMismatch),
+            }
+        }
```
I would probably extract this as a free function. I also think this could take a `K: KeyPair` instead of taking the `pubkey` directly. You could even go all in on this and abstract towards an `impl Fn(&[u8]) -> Option<impl KeyPair>` to reduce duplication more.

(Style nit: despite existing code, I think `pub_key` would be more idiomatic spelling in Rust than `pubkey`. It's also not obvious to me what `_tbv` means. Maybe something more like `cert_pub_key` and `pair_pub_key`?)

cpu replied:
Reworked :+1:

djc commented on src/end_entity.rs:
```diff
@@ -185,4 +194,97 @@ impl<'a> EndEntityCert<'a> {
+    extractor: impl Fn(&[u8]) -> Option<K>,
+    cert_pub_key: &[u8],
+) -> Option<Result<(), Error>> {
+    match extractor(private_key_der) {
```
You can replace the outer `match` with `let key_pair = extractor(private_key_der)?;`. Also consider extracting the `Some` from the inner match arms.

cpu commented on src/end_entity.rs:
```diff
@@ -185,4 +194,97 @@ impl<'a> EndEntityCert<'a> {
+    ///   * PKCS8v2 (Ed25519 only)
+    ///   * PKCS1 (RSA only)
+    ///   * Sec1 (ECDSA only)
+    pub fn verify_private_key(&self, private_key_der: &[u8]) -> Result<(), Error> {
```
Those are compelling advantages :+1: In retrospect I was probably overfitting [Brian's suggestion](https://github.com/rustls/rustls/issues/618#issuecomment-817009479) on Rustls#618 by jumping straight to this design.

cpu commented on src/end_entity.rs:
```diff
@@ -185,4 +194,97 @@ impl<'a> EndEntityCert<'a> {
+    cert_pub_key: &[u8],
+) -> Option<Result<(), Error>> {
+    match extractor(private_key_der) {
+        Some(keypair) => match cert_pub_key == keypair.public_key().as_ref() {
```
That's a good point. We should also be asserting the key types match somewhere in this logic. I'll give this more attention in the reworked formulation. Thanks for calling this out as a potential pitfall.
